rfc/rfc-88/rfc-88.md

```diff
@@ -0,0 +1,601 @@
+        
+        public Builder addField(Field field) {..}
+        ...
+        public Schema build() {
+            return new Schema(fields, versionId, maxColumnId, recordKeys, partitionKeys);
+        }
+    }
+}
+
+/**
+ * Definition of the type of record in Hudi.
+ */
+@PublicAPIClass
+public class RecordType implements Type {
+    // name is necessary to provide for lossless conversion b/w Avro and Schema
+    private final String name;
+    private final Field[] fields;
+    
+    public RecordType(List<Field> fields) {}
+    public RecordType(String name, List<Field> fields) {}
+    
+    public List<Field> fields() {}
+    public Field field(String name) {}
+    public Field field(int id) {}
+}
+
+/**
+ * Definition of a column/field in RecordType.
+ */
+@PublicAPIClass
+public static class Field implements Serializable {
+    private final boolean isOptional;
+    private final int id;
+    private final String name;
+    private final Type type;
+    private final String doc;
+    private final Object defaultValue;
+    private NestedField() {}
+    
+    public static class Builder {
+        private final String name;
+        ...
+        private final Object defaultValue;
+        
+        public Builder(String name) {
+            this.name = name;
+        }
+        
+        public Builder withId(int id) {..}
+        ...
+        public NestedField build() {
+          return new Field(isOptional, id, name, type, doc, defaultValue);
+        }
+    }
+}
+```
+For the first stated goal, the following changes are proposed:
+* Substitute avro schema with `Schema` in commit metadata.
+  - Persistence: SCHEMA_KEY: Avro schema string -> `Schema` string
+  - Resolving compatibility: firstly try resolving as `Schema`, and fallback to Avro schema for compatibility.
+* Substitute Avro schema with `Schema` in reader/writer path of all engines.
+  - Replace schemas in configuration.
+  - Replace schemas in reader/writer building process.
+* Substitute Avro schema with `Schema` in log block.
+
+### Abstraction of Writers/Readers
+In the lake architecture, abstracting low-level file readers and writers is a crucial design choice. With a well defined 
+reader/writer abstraction, we can make the system more flexible, extendible, and reduce the efforts required for development 
+and maintenance. Furthermore, the reading/writing behaviors among different engines can be unified, as well as sharing the performance optimization.
+#### Writer Abstraction
+![writer_abstract](writer_abstract.png)
+The proposed abstraction of writer  is shown in the diagram. It's generally divided into two levels:
+* FileWriter: FileWriter is the basic interface of writers for files in different formats. It's composed of a StructWriter and related writing interface provided by the underlying formats, such as ParquetWriter  and WriteSupport  for Parquet format.
+    ```java
+    public interface FileWriter<R> extends Closeable {
+        void write(R value);
+    }
```

Comment:
As @zhangyue19921010 mentioned, we implemented the same approach in our production, we write RowData directly to parquet files.

So here is the main question - Where to will FileWriter write data? If it will be file represented by StoragePath, we will have problems with writing log blocks for MOR table. For writing log block firstly we need to create data content in memory, then calculate size of this data, then write size and data itself in log block (`HoodieLogFormatWriter`).

Problem with unavailability of in memory write was introduced by StoragePath [HUDI-6243](https://issues.apache.org/jira/browse/HUDI-6243) abstraction, but we need to create a workaround here. Also FileWriter will be more testable in this case.